thetis/callback.py

```diff
@@ -861,3 +870,151 @@ def message_str(self, *args):
                 fieldname, self.location_name, minval, maxval)
         out = out[:-1]  # suppress last line break
         return out
+
+
+class TransectCallback(DiagnosticCallback):
+    """
+    Extract a vertical transect of a 3D field at a given (x,y) locations.
+
+    Only for the 3D model.
+    """
+    name = 'transect'
+    variable_names = ['z_coord', 'value']
+
+    def __init__(self, solver_obj, fieldnames, x, y,
+                 location_name,
+                 n_points_z=48,
+                 z_min=None, z_max=None,
+                 outputdir=None, export_to_hdf5=True,
+                 append_to_log=True):
+        """
+        :arg solver_obj: Thetis :class:`FlowSolver` object
+        :arg fieldnames: List of fields to extract
+        :arg x, y: location coordinates in model coordinate system.
+        :arg location_name: Unique name for this location. This
+            determines the name of the output h5 file (prefixed with
+            `diagnostic_`).
+        :arg int n_points_z: Number of points along the vertical axis. The 3D
+            field will be interpolated on these points, ranging from the bottom
+            to the (time dependent) free surface.
+        :kwarg float z_min, zmax: Force min/max value of z coordinate extent.
+            By default, transect will cover entire depth from bed to surface.
+        :kwarg str outputdir: Custom directory where hdf5 files will be stored.
+            By default solver's output directory is used.
+        :kwarg bool export_to_hdf5: If True, diagnostics will be stored in hdf5
+            format
+        :kwarg bool append_to_log: If True, callback output messages will be
+            printed in log
+        """
```

Comment:
Might be nice to explicitly say it only prints a min max of each field - as I would be worried as a user this would dump a whole lot of data in the log.